tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,763 @@
+	}
+
+	if t.Int64Val != nil {
+		return t.Int64Val, nil
+	}
+
+	data := t.Data()
+	res = make([]int64, len(data)/cBytesInt64)
+	for i := range res {
+		res[i] = int64(binary.LittleEndian.Uint64(data[i*cBytesInt64 : (i+1)*cBytesInt64]))
+	}
+	t.Int64Val = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// AsBool returns the content of the tensor as a slice of bool if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DtBool
+func (t *Tensor) AsBool() (res []bool, err error) {
+	if DtBool != t.DataType() {
+		err = &ErrInvalidTensorType{
+			tensorType:   t.DataType(),
+			expectedType: DtBool,
+		}
+		return
+	}
+
+	if t.BoolVal != nil {
+		return t.BoolVal, nil
+	}
+
+	data := t.Data()
+	res = make([]bool, len(data))
+	for i, v := range data {
+		res[i] = v == 1
+	}
+	t.BoolVal = res
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return
+}
+
+// GetVal resturns the value of the element contained in the specified position
+// on the tensor, Ex: GetVal(1, 2, 3) is equivalent to data[1][2][3] on a
+// multidimensional array.
+// This method could return an error in case of a wrong specified number of
+// dimensions or a dimesions out of range.
+func (t *Tensor) GetVal(d ...int) (val interface{}, err error) {
+	if len(d) != t.NumDims() {
+		err = &ErrDimsOutOfTensorRange{
+			tensorDim: t.NumDims(),
+			specDims:  len(d),
+		}
+		return
+	}
+
+	pos := 0
+	if t.dimWeights != nil {
+		for i, w := range t.dimWeights {
+			pos += d[i] * w
+		}
+	} else {
+		t.dimWeights = make([]int, len(d))
+		pos = d[len(d)-1]
+		if pos >= t.Dim(len(d)-1) {
+			err = &ErrIndexOutOfRange{
+				dim:       len(d) - 1,
+				index:     pos,
+				dimsRange: t.Dim(len(d) - 1),
+			}
+			return
+		}
+		t.dimWeights[len(d)-1] = 1
+
+		lastWeight := 0
+		for i := len(d) - 2; i >= 0; i-- {
+			lastWeight += t.Dim(i + 1)
+			t.dimWeights[i] = lastWeight
+			pos += d[i] * lastWeight
+
+			if d[i] >= t.Dim(i) {
+				err = &ErrIndexOutOfRange{
+					dim:       i,
+					index:     pos,
+					dimsRange: t.Dim(i),
+				}
+				return
+			}
+		}
+	}
+
+	switch t.DataType() {
+	case DtFloat:
+		vals, _ := t.AsFloat32()
+		val = vals[pos]
+	case DtDouble:
+		vals, _ := t.AsFloat64()
+		val = vals[pos]
+	case DtInt8, DtInt16, DtInt32, DtUint8:
+		vals, _ := t.AsInt32()
+		val = vals[pos]
+	case DtInt64:
+		vals, _ := t.AsInt64()
+		val = vals[pos]
+	case DtBool:
+		vals, _ := t.AsBool()
+		val = vals[pos]
+	case DtString:
+		vals, _ := t.AsStr()
+		val = vals[pos]
+	default:
+		err = &ErrTensorTypeNotSupported{
+			tensotType: t.DataType(),
+		}
+		return
+	}
+
+	return
+}
+
+// setCMemAsAlreadyRelease The C allocated memory was already released from C.
+func (t *Tensor) setCMemAsAlreadyRelease() {
+	t.memReleased = true
+}
+
+// FreeAllocMem Method used telease the C allocated memory for this tensor.
+func (t *Tensor) FreeAllocMem() {
+	// We can't clean the tensor here in case of it had been  used as in
```

Comment:
in case of it -> in case it
as in input -> as an input
since on ->  , because in
is cleaning -> cleans
every execution what -> every execution.  This
typo: 'a a' -> a